Improve handling of URI fragment identifiers

AlbertHilb opened the pull request:
Fixes #5599, #4190.

Supersedes #5613.

This adds `handleFragment` method to `IDocumentWidget` and implements fragment handling for markdown and PDF documents, and notebooks.

Cheers

ian-r-rose commented on packages/rendermime/src/renderers.ts:
```diff
@@ -414,6 +423,11 @@ export namespace renderMarkdown {
+    /**
+     * The `id` of the element should be scrolled into view.
+     */
+    fragment: string | null;
```
Can we make this an optional parameter instead of `string | null`?

ian-r-rose replied:
I think the `fragment` parameter can be moved into the `renderedHTML` options, since they are both dealing with rendered DOM content. Then we can use fragments for both html content and markdown content.

AlbertHilb replied:
> Can we make this an optional parameter instead of `string | null`?

Yes. I made it `string | null` for coherence with the other optional interface members.

> I think the `fragment` parameter can be moved into the `renderedHTML` options

It would be an excellent idea if `renderMarkdown` used `renderHTML` to put the HTML code into the host element. But, if I'm right, [it does not][1] at the moment.

[1]: https://github.com/jupyterlab/jupyterlab/blob/ff47903e8adfc4cd13a688734355e18b48fb2dbe/packages/rendermime/src/renderers.ts#L355

ian-r-rose commented on packages/rendermime/src/renderers.ts:
```diff
@@ -398,6 +399,14 @@ export function renderMarkdown(
         latexTypesetter.typeset(host);
       }
+    })
+    .then(() => {
```
I think this `.then` clause can be consolidated with the one above, since it resolves more or less immediately (the typesetting is asynchronous, but we don't wait on it).

AlbertHilb replied:
OK. If we don't wait for LaTeX typesetting, we can join the two clauses.

ian-r-rose commented on packages/cells/src/widget.ts:
```diff
@@ -864,6 +864,7 @@ export class MarkdownCell extends Cell {
     this._updateRenderedInput().then(() => {
       this._ready.resolve(void 0);
     });
+    this.renderInput(this._renderer);
```
What does this change accomplish?

AlbertHilb replied:
In order to scroll into view the fragment we have to wait cell output is inserted into DOM.
However `_updateRenderedInput()` just generates the output without injecting it into DOM.
So, without that change, the `ready` promise resolves too early for our needs.

ian-r-rose commented on packages/docregistry/src/mimedocument.ts:
```diff
@@ -105,6 +113,7 @@ export class MimeContent extends Widget {
   protected onUpdateRequest(msg: Message): void {
     if (this._context.isReady) {
       this._render();
+      this._fragment = null;
```
Any particular reason to null this out (not that I see a problem with it)?

AlbertHilb replied:
`_fragment` value makes sense only until the document is rendered the first time after that value is set. Later the user can scroll the page, for example, and he doesn't want it jumps back to the old position on following rerenderings.

AlbertHilb commented:
Hi @ian-r-rose. I tested this new functionality on Firefox and Chrome for Ubuntu. Everything seems to work fine, except for a (minor) issue with the PDF viewer integrated in Chrome: it recognizes fragment of type `#page=x`, but not of type `search=x`.

blink1073 commented:
Caught by the linter:

`ERROR: packages/docregistry/src/default.ts[469, 42]: block is empty`.

You can put a `/* no-op */` comment in there to avoid the lint error.

AlbertHilb commented:
Comment added.

ian-r-rose commented on packages/rendermime/src/renderers.ts:
```diff
@@ -398,6 +399,12 @@ export function renderMarkdown(
         latexTypesetter.typeset(host);
       }
+      if (fragment) {
+        const el = host.querySelector(fragment);
```
Since these are user-settable, I found it pretty easy to accidentally enter an invalid fragment, which threw an error. Maybe this block could be in a try/except statement to catch those (possibly with a console warning)?

AlbertHilb replied:
I agree.

ian-r-rose commented on packages/notebook/src/widget.ts:
```diff
@@ -1310,6 +1321,13 @@ export class Notebook extends StaticNotebook {
       activeCell.addClass(OTHER_SELECTED_CLASS);
     }
+    if (this._fragment) {
+      const el = this.node.querySelector(this._fragment);
```
Same here as below.

AlbertHilb replied:
I agree again.

ian-r-rose commented on packages/docregistry/src/mimedocument.ts:
```diff
@@ -133,7 +142,11 @@ export class MimeContent extends Widget {
+    let mimeModel = new MimeModel({
+      data,
+      callback: this._changeCallback,
+      metadata: { fragment: this._fragment }
```
I think it would be useful to note in the documentation for `MimeModel` that `fragment` may be set in the metadata, and that renderers may choose to respect that (or not).

AlbertHilb replied:
It's useful indeed.

ian-r-rose commented:
I especially love the handling of PDF fragments.

AlbertHilb commented:
Hi @ian-r-rose.

I rebased this PR to remove conflicts introduced by #5674.
Following your suggestions I added console warning for malformed URI fragments and some note about `fragment` attribute in `IMimeModel` `metadata` documentation.
Finally, I moved the markdown fragment handling logic from `rendermime/renderers` to `rendermime/widgets` and I reshaped it in order to share code among all the widgets based on `RenderedHTMLCommon`.

AlbertHilb commented on packages/rendermime/src/widgets.ts:
```diff
@@ -242,6 +242,7 @@ export class RenderedMarkdown extends RenderedHTMLCommon {
   render(model: IRenderMime.IMimeModel): Promise<void> {
     return renderers.renderMarkdown({
       host: this.node,
+      fragment: String(model.metadata.fragment),
```
You are totally right.

ian-r-rose commented:
I am seeing some problems with how Lab is handling fragment identifiers (especially with special characters `':'` and additional HTML markup), but that is not your fault, so let's move forwards with this and I'll add a new issue for that.
